apps/site/lib/site_web/controllers/schedule/line_controller.ex

```diff
@@ -43,28 +44,75 @@ defmodule SiteWeb.ScheduleController.LineController do
   @spec get_schedules(binary, any, any) :: %{by_trip: map, trip_order: [any]}
   def get_schedules(route_id, date, direction_id) do
     services =
-      Enum.map(
-        Schedules.Repo.by_route_ids([route_id], date: date, direction_id: direction_id),
-        &Map.update!(&1, :route, fn route -> Route.to_json_safe(route) end)
-      )
+      [route_id]
+      |> Schedules.Repo.by_route_ids(date: date, direction_id: direction_id)
+      |> Enum.map(&Map.update!(&1, :route, fn route -> Route.to_json_safe(route) end))
+
+    ordered_trips = services |> Enum.sort_by(& &1.time) |> Enum.map(& &1.trip.id) |> Enum.uniq()
 
     services_by_trip =
       services
-      |> Enum.map(&Map.update!(&1, :time, fn time -> Timex.format!(time, "{0h12}:{m} {AM}") end))
       |> Enum.group_by(& &1.trip.id)
 
-    ordered_trips = services |> Enum.sort_by(& &1.time) |> Enum.map(& &1.trip.id) |> Enum.uniq()
-    %{by_trip: services_by_trip, trip_order: ordered_trips}
+    services_by_trip_with_fare =
+      services_by_trip
+      |> Stream.map(fn {trip_id, service} -> {trip_id, fares_for_service(service)} end)
+      |> Stream.map(fn {trip_id, service} -> {trip_id, duration_for_service(service)} end)
+      |> Stream.map(fn {trip_id, service} -> {trip_id, formatted_time(service)} end)
+      |> Enum.into(%{})
+
+    %{by_trip: services_by_trip_with_fare, trip_order: ordered_trips}
+  end
+
+  def fares_for_service(schedules) do
+    origin = List.first(schedules)
+
+    schedules
+    |> Enum.map(
+      &Map.merge(
+        &1,
+        fares_for_service(origin.route, origin.stop.id, &1.stop.id)
+      )
+    )
+  end
+
+  def duration_for_service(schedules) do
+    first = List.first(schedules).time
+    last = List.last(schedules).time
+    %{schedules: schedules, duration: Timex.diff(last, first, :minutes)}
+  end
+
+  def formatted_time(%{schedules: schedules, duration: duration}) do
+    time_formatted_schedules =
+      schedules
+      |> Enum.map(&Map.update!(&1, :time, fn time -> Timex.format!(time, "{0h12}:{m} {AM}") end))
```

Comment:
may make sense to deal w/ date formatting on client since we generally know `Timex` is slow

Reply:
I'm not sure we should make it a hard fast rule to do date formatting on the client.

We already use this function for trip duration on schedule pages (see TripInfo module) that have trips listed.

Doing the trip duration clientside correctly would probably cause me to bring in a new library. So bundle size is a concern in this case. In other cases, we'd have to handle timezone correctly (usually EST but parsing dates clientside needs to be correct on that) which is also often easier with a library.

I think because this is just comparing the first -> last date from every trip, it's okay. I will add timing to get a better idea, and see if DateTime can do it instead.

Reply:
Phil and I looked at this last time we met to review performance... and it is surprisingly slow :/

Reply:
This code to get the trip duration was not part of the React work previously, so I didn't think you would have checked the old EEX templates for this.

Reply:
Ah I see I misplaced where you placed this comment. I did introduce a Timex.diff call somewhere else.

This code is exactly as it was, just moved around, not new, so I won't change it in this PR as this PR doesn't address the performance stuff.